sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SyntaxTrackers/CSharpObjectInitializationTracker.cs

```diff
@@ -0,0 +1,263 @@
+        {
+            if (expression == null)
+            {
+                return false;
+            }
+            if (expression.IsKind(SyntaxKind.NullLiteralExpression))
+            {
+                return true;
+            }
+            if (semanticModel.GetConstantValue(expression).Value is { } constantValue)
+            {
+                return IsAllowedConstantValue(constantValue);
+            }
+            if (semanticModel.GetSymbolInfo(expression).Symbol is { } symbol)
+            {
+                return IsAllowedObject(symbol);
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Verifies that the properties are assigned with allowed values. Otherwise, verifies the constructor invocation.
+        /// </summary>
+        /// There are multiple ways of verifying:
+        /// - implicit constructor with property setting
+        ///   var x = new X { Prop = new Y() };
+        /// - constructor with passing a list of arguments, where we care about a specific argument
+        /// <remarks>
+        /// Currently we do not handle the situation with default and named arguments.
+        /// </remarks>
+        private bool ObjectCreatedWithAllowedValue(ObjectCreationExpressionSyntax objectCreation, SemanticModel semanticModel)
+        {
+            var trackedPropertyAssignments = GetInitializerExpressions(objectCreation.Initializer)
+                .OfType<AssignmentExpressionSyntax>()
+                .Where(assignment => IsTrackedPropertyName(assignment.Left))
+                .ToList();
+
+            if (trackedPropertyAssignments.Count != 0)
+            {
+                return trackedPropertyAssignments.All(assignment => IsAllowedValue(assignment.Right, semanticModel));
+            }
+
+            var argumentList = objectCreation.ArgumentList;
+            if (trackedConstructorArgumentIndex != -1)
+            {
+                return argumentList == null ||
+                    argumentList.Arguments.Count != trackedConstructorArgumentIndex + 1 ||
+                    IsAllowedValue(argumentList.Arguments[trackedConstructorArgumentIndex].Expression, semanticModel);
+            }
+
+            // if no tracked properties are being explicitly set or passed as arguments, the constructor may still be safe by design
+            return constructorIsSafe;
+        }
+
+        /// <summary>
+        /// Tests the name of the property that has to be set with an allowed value in order for the object to be initialized correctly.
+        /// </summary>
+        /// <returns>True when <paramref name="propertyName"/> is the name of a relevant property, otherwise false.</returns>
+        private bool IsTrackedPropertyName(string propertyName) => isTrackedPropertyName(propertyName);
+
+        /// <summary>
+        /// Tests if the <paramref name="expression"/> is <see cref="TrackedPropertyName"/>.
```

Comment:
I think the comment is out of sync. We don't have a `TrackedPropertyName` anymore.

Reply:
indeed, and the comment isn't useful at all - the method name is clear and the body is short